src/main/kotlin/com/mdrobnak/lalrpop/codeInsight/LpExpressionTypeInlayHintProvider.kt

```diff
@@ -0,0 +1,126 @@
+            // them, which this preview is not able to provide
+            grammar;
+            
+            Comma<Rule> = (<Rule> ",")+ Rule?;
+            
+            FunctionArgs = Comma<Expression>;
+            
+            Expression = r"[0-9]+";
+        """.trimIndent()
+
+    override fun createConfigurable(settings: Settings): ImmediateConfigurable = object : ImmediateConfigurable {
+        override val cases: List<Case>
+            get() = listOf(
+                Case("Show for nonterminals", "nonterminals", settings::showForNonterminals),
+                Case("Show for symbols", "symbols", settings::showForSymbols),
+                Case("Ignore unselected symbols", "unselected_symbols", settings::ignoreUnselectedSymbols),
+                Case(
+                    "Only show for symbols where the alternative has an action",
+                    "alternative_action_symbols",
+                    settings::alternativeActionSymbols
+                )
+            )
+
+        override fun createComponent(listener: ChangeListener): JComponent = JPanel()
+    }
+
+    override fun createSettings(): Settings = Settings()
+
+    override fun getCollectorFor(
+        file: PsiFile,
+        editor: Editor,
+        settings: Settings,
+        sink: InlayHintsSink
+    ): InlayHintsCollector = object : FactoryInlayHintsCollector(editor) {
+        val typeHintsPresentationFactory = RsTypeHintsPresentationFactory(factory, false)
+
+        override fun collect(element: PsiElement, editor: Editor, sink: InlayHintsSink): Boolean {
+            if (file.project.service<DumbService>().isDumb) return true
+
+            if (element is LpNonterminal) presentTypeForNonterminal(element)
+            if (element is LpAlternative) presentTypesForSymbols(element)
+
+            return true
+        }
+
+        private fun presentTypesForSymbols(alternative: LpAlternative) {
+            if (!settings.showForSymbols) return
+            if (settings.alternativeActionSymbols && alternative.action == null) return
+            (if (settings.ignoreUnselectedSymbols) alternative.selected else alternative.symbolList).forEach(this::presentTypeForSymbol)
+        }
+
+        private fun presentType(nonterminal: LpNonterminal, element: LpResolveType, anchor: PsiElement) {
+            val type =
+                element.getContextAndResolveType(LpMacroArguments.identity(nonterminal.nonterminalName.nonterminalParams))
+
+            val modDecl = findModuleDefinition(element.project, element.containingFile) ?: return
+
+            val rustType = type.lalrpopRustType(
+                element.project,
+                element.containingFile.importCode,
+                nonterminal,
+                modDecl
+            ) ?: return
+
+            val inlayPresentation = typeHintsPresentationFactory.typeHint(rustType)
+
+            @Suppress("Deprecation")
+            sink.addInlineElement(anchor.endOffset, false, inlayPresentation)
```

Comment:
This particular form of `addInlineElement` is experimental and deprecated, i.e. it is definitely not safe to rely on. Use the suggested alternative which is at least not deprecated:
```suggestion
            sink.addInlineElement(anchor.endOffset, false, inlayPresentation, false)
```

Reply:
The rust plugin relies on it :thinking:
https://github.com/intellij-rust/intellij-rust/blob/master/src/main/kotlin/org/rust/ide/hints/type/RsInlayTypeHintsProvider.kt#L171-L173

Reply:
Hm, the comment mentions compatibility with 2020.2. I just tried running this under 2020.2 and it does fail to compile due to a missing method. I guess like intellij-rust we'll have to keep using the deprecated version until we drop 2020.2.